# SOLUTIONS AND EXPLANATIONS FOR LESSON 2: PROBABILITY AND STATISTICS FOR AI (21/6/2026) HOMEWORK

## PREPARE DATASET 

The Iris dataset is short for Iris flower dataset which stores physical sizes of different types of Iris flowers.

If we import the dataset from seaborn lib

In [43]:
import pandas as pd
import seaborn as sns

Iris_sns = sns.load_dataset("iris")

print(Iris_sns.head())
print(Iris_sns.shape)

   sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
(150, 5)


If we import the dataset from sklearn lib

In [44]:
from sklearn.datasets import load_iris

iris_raw = load_iris()

print(iris_raw)

{'data': array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
     

As we can see, sklearn lib store iris data as Numpy array (a dictionary to be specific), so to visualize better we can use pandas to format it to be similar to seaborn.

In [45]:
iris = pd.DataFrame(iris_raw.data, columns = iris_raw.feature_names)
print(iris)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Reading from the sklearn description about iris dataset, we can learn that iris contains 150 flowers classified into 3 types of iris.

## EXERCISE 1: DESCRIPTIVE STATISTICS & CHARACTERISTICS

In [46]:
print(Iris_sns.shape)
print(Iris_sns.dtypes)

# first 5 rows
print(Iris_sns.head(5)) # or Iris_sns.iloc[:5, :]

(150, 5)
sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
species             str
dtype: object
   sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa


Iris dataset has 150 rows and 5 columns:
- First 4 columns (sepal_length/width and petal_length/width) contains float-type numbers
- The last column (species) contains string (types of Iris)

### MEAN, MEDIAN, MODE, VAR, STD, MIN, MAX, Q123, IQR

Pandas lib has a built-in function that can calculate most of these for us.

In [47]:
print(Iris_sns.describe())

       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000


Or we can calculate by each column.

In [65]:
mean1 = Iris_sns["sepal_length"].mean()
print(f"mean of sepal_length: {mean1}")

mean2 = Iris_sns["sepal_width"].mean()
print(f"mean of sepal_width: {mean2}")

mean3 = Iris_sns["petal_length"].mean()
print(f"mean of petal_length: {mean3}")

mean4 = Iris_sns["petal_width"].mean()
print(f"mean of petal_width: {mean4}")

# or we can calculate mean by formula
def mean(query):
    total = 0
    for i in range(len(Iris_sns[query])):
        total += Iris_sns[query][i]
    return (f"mean of {query} is: {total/len(Iris_sns[query])}")

mean("sepal_length")

mean of sepal_length: 5.843333333333334
mean of sepal_width: 3.0573333333333337
mean of petal_length: 3.7580000000000005
mean of petal_width: 1.1993333333333336


'mean of sepal_length is: 5.843333333333335'

In [ ]:
median1 = Iris_sns["sepal_length"].median()
print(f"median of sepal_length: {median1}")

median2 = Iris_sns["sepal_width"].median()
print(f"median of sepal_width: {median2}")

median3 = Iris_sns["petal_length"].median()
print(f"median of petal_length: {median3}")

median4 = Iris_sns["petal_width"].median()
print(f"median of petal_width: {median4}")

#or we can calculate median by formula
def median(query):
    if len(Iris_sns[query]) % 2 == 0:
        pos = len(Iris_sns[query]) / 2 - 1
        return (f"median of {query}")

median of sepal_length: 5.8
median of sepal_width: 3.0
median of petal_length: 4.35
median of petal_width: 1.3


Note: median = Q2, lies at the middle of the data (50%).

In [ ]:
mode1 = Iris_sns["sepal_length"].mode()[0]
print(f"mode of sepal_length: {mode1}")

mode2 = Iris_sns["sepal_width"].mode()[0]
print(f"mode of sepal_width: {mode2}")

mode3 = Iris_sns["petal_length"].mode()[0]
print(f"mode of petal_length: {mode3}")

mode4 = Iris_sns["petal_width"].mode()[0]
print(f"mode of petal_width: {mode4}")

mode = Iris_sns["sepal_length"].mode()
print(mode)

mode of sepal_length: 5.0
mode of sepal_width: 3.0
mode of petal_length: 1.4
mode of petal_width: 0.2
0    5.0
Name: sepal_length, dtype: float64


mode() outputs a pandas array (dictionary) has matches the actual mode to an index. In this case is 0 as there's only one mode.  
But if the data has more than 2 modes the index will continue 0 -> 1 -> 2, so we have to call mode()[0] not mode()[1]